--- a/linux-user/elfload.c	2020-07-13 20:10:37.776374566 -0700
+++ b/linux-user/elfload.c	2020-07-13 20:11:51.794957015 -0700
@@ -2667,6 +2667,11 @@
     char *elf_interpreter = NULL;
     char *scratch;
 
+    memset(&interp_info, 0, sizeof(interp_info));
+#ifdef TARGET_MIPS
+    interp_info.fp_abi = MIPS_ABI_FP_UNKNOWN;
+#endif
+
     info->start_mmap = (abi_ulong)ELF_START_MMAP;
 
     load_elf_image(bprm->filename, bprm->fd, info,

