--- pngrutil.c.orig	2014-06-12 03:35:16.000000000 +0200
+++ pngrutil.c	2014-07-01 05:08:31.000000000 +0200
@@ -268,7 +268,11 @@
    if (need_crc != 0)
    {
       crc = png_get_uint_32(crc_bytes);
-      return ((int)(crc != png_ptr->crc));
+
+      if (crc != png_ptr->crc)
+        fprintf(stderr, "NOTE: CRC in the file is 0x%08x, change to 0x%08x\n", crc, png_ptr->crc);
+
+      return ((int)(1 != 1));
    }
 
    else
